docs/source/architecture.rst

```diff
@@ -1,19 +1,740 @@
+
+Since Insolar sets apart functionality using :ref:`node roles <multi_role_nodes>`, it has two sets of execution & validation procedures: **virtual** and **material**.
+
+.. _virtual_execution_validation:
+
+Virtual Execution & Validation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Nodes with :ref:`virtual static roles <virtual>` carry out **virtual** execution & validation:
+
+#. The network selects (determines based on :term:`entropy <pulse>`) a specific virtual node to become a :ref:`virtual executor <dynamic_roles>`. Upon receiving the request, the executor:
+
+   #. Registers the request within the current :term:`pulse`.
+
+      In case the request refers to a 'busy' :term:`object <object>`, execution may be delayed and handed over to another executor due to state locks or multiple updates. Moreover, multiple requests can be executed within the same pulse when opportunistic execution/validation is allowed by the caller or by the called object.
+
+   #. Executes the request on the :term:`object <object>` (contract).
+   #. Collects the results of outbound calls.
+   #. Provides :term:`lifeline <lifeline>` and :term:`sideline <sideline>` updates for validation by other nodes.
+
+#. Once the executor’s status expires, the network selects :ref:`virtual validators <dynamic_roles>` from the list of active :ref:`virtual nodes <virtual>` on a new :term:`pulse <pulse>`, meaning executors cannot predict which nodes will validate transactions, thereby avoiding a collusion scenario. 
+
+#. Each virtual validator:
+
+   #. Checks that the request is legitimate.
+   #. Executes the request on the :term:`object <object>` (contract) a second time.
+   #. Checks that the request returns the same response given the :ref:`same arguments <contract_determinism>`.
+   #. Checks that the request performs the same outbound calls.
+
+#. Lastly, the outbound calls validation is stacked into a single validation round as validators use signed results collected by previous executors.
+
+A single virtual executor can execute long requests that span several pulses. To do this, the virtual node that started the execution asks current executors in each pulse for tokens that give the execution permission.
+
+.. _material_execution_validation:
+
+Material Execution & Validation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Nodes with :ref:`light material static roles <virtual>` carry out **material** execution & validation:
+
+#. The network selects (determines based on :term:`entropy <pulse>`) a specific light material node to become a :ref:`material executor <dynamic_roles>`. Upon receiving data requests from the virtual executor in the current :term:`pulse <pulse>`, the material executor:
```

Comment:
it is better to avoid saying just "material executor" in the doc please, but be more specific please. hence "light material executor" here